src/cdk-experimental/selection/selection.ts

```diff
@@ -0,0 +1,223 @@
+      true,
+      this.trackBy
+    );
+
+    if (this._selectionModel.change) {
+      this._selectionModel.change.subscribe(e => {
+        this.selectedItems = this._selectionModel.selected;
+        this.selectionChange.emit(e);
+      });
+    }
+  }
+
+  /** Register the toggle directive. */
+  registerToggle(toggle) {
+    this._toggleDirectives.push(toggle);
+  }
+
+  /** Deregister the toggle directivet. */
+  deregisterToggle(toggle) {
+    const toggleIndex = this._toggleDirectives.indexOf(toggle);
+    if (toggleIndex > -1) {
+      this._toggleDirectives.splice(toggleIndex, 1);
+    }
+  }
+
+  /** Toggle the selection of a item. */
+  toggle(toggle) {
+    if (this.disabled) {
+      return;
+    }
+
+    if (Array.isArray(toggle.model)) {
+      if (this.mode === 'multiple' &&
+          (this.maxSelections === undefined ||
+          (toggle.model.length + this._selectionModel.selected.length) < this.maxSelections)) {
+        this._selectionModel.clear();
+        this._selectionModel.toggle(...toggle.model);
+      }
+    } else {
+
+      this._setSelections(toggle);
+    }
+  }
+
+  /** Set the user selection for the element. */
+  setTextSelection(type: string) {
+    const elm = this._elementRef.nativeElement;
+    elm.style.userSelect = type;
+    elm.style.webkitUserSelect = type;
+    elm.style.MozUserSelect = type;
+  }
+
+  /** Set the selection state given a toggle directive and its activated options. */
+  private _setSelections(toggle) {
+    const ctrls = this._getSortedSet();
+    const index = this._getToggleIndex(ctrls, toggle);
+    const { modifier, model } = toggle;
+    const selections = this._selectionModel.selected;
+
+    if (this.mode === 'multiple' && modifier === 'shift') {
+      if (index > -1) {
+        this._selectRange(ctrls, this._previousSelectedIndex, index);
+      }
+    } else if (this.mode === 'multiple') {
+      const isSelected = this._selectionModel.isSelected(model);
+      if (this.requireModifier && modifier !== 'meta') {
+        this._selectionModel.clear();
+        if (!isSelected || !this.deselectable) {
+          this._selectionModel.select(model);
+        }
+      } else {
+        if (!isSelected) {
+          if (this.maxSelections === undefined || selections.length < this.maxSelections) {
+            this._selectionModel.select(model);
+          }
+        } else if (this.deselectable || (!this.deselectable && selections.length > 1)) {
+          this._selectionModel.deselect(model);
+        }
+      }
+    } else {
+      const isSelected = this._selectionModel.isSelected(model);
+      if (!isSelected) {
+        this._selectionModel.select(model);
+      } else if(this.deselectable) {
+        this._selectionModel.deselect(model);
+      }
+    }
+
+    this._previousSelectedIndex = index;
+  }
+
+  /** Selects a collection of values between two indexes. */
+  private _selectRange(ctrls: any[], sourceIndex: number, targetIndex: number) {
```

Comment:
Add `@param` JsDoc to clarify whether they're inclusive/exclusive, whether they need to be in a certain order, and what "source" and "target" represent?